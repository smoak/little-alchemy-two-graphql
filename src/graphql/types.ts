export interface Node {
  readonly id: string;
}

export interface ItemCombination {
  readonly source: Item;
  readonly target: Item;
}

export interface Item extends Node {
  readonly name: string;
  readonly myths: boolean;
  readonly creates: ItemCombination[];
}

export interface Query {
  readonly items: Item[];
}
