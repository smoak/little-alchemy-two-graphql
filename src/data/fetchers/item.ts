import low from 'lowdb';
import FileSync from 'lowdb/adapters/FileSync';

const adapter = new FileSync<DatabaseSchema>('db.json');
const db = low(adapter);

interface DatabaseSchema {
  readonly items: DatabaseItems;
}

export interface DatabaseItem {
  readonly myths: boolean;
  readonly makes: Record<string, string>;
  readonly combinations: string[][];
}

type DatabaseItems = Record<string, DatabaseItem>;

type GetItems = () => DatabaseItems;
export const getItems: GetItems = () => db.get('items').value();

type GetItemByName = (name: string) => DatabaseItem | undefined;
export const getItemByName: GetItemByName = name => db.get('items').value()[name];

type GetItemNames = () => string[];
export const getItemNames: GetItemNames = () => db.get('items').keys().value();
