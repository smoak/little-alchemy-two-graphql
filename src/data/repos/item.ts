import Fuse from 'fuse.js';

import { DatabaseItem, getItemByName, getItemNames, getItems } from '../fetchers/item';

export interface Item {
  readonly name: string;
  readonly myths: boolean;
  readonly creates: ItemCombination[];
  readonly combinations: ItemCombination[];
  readonly imageUrl: string;
}

export interface ItemCombination {
  readonly source: string;
  readonly target: string;
}

const fromCombinations = (combinations: DatabaseItem['combinations']): ItemCombination[] =>
  combinations.map(combo => ({ source: combo[0], target: combo[1] }));

const fromItemWithName = (name: string, item: DatabaseItem): Item => ({
  name,
  imageUrl: item.image_url,
  myths: item.myths,
  creates: item.makes,
  combinations: fromCombinations(item.combinations),
});

type FindById = (id: string) => Item;
export const findById: FindById = id => {
  const item = getItemByName(id);

  if (!item) {
    throw new Error(`Could not find item with id: ${id}`);
  }

  return fromItemWithName(id, item);
};

type FindAll = () => Item[];
export const findAll: FindAll = () => {
  const items = getItems();

  return Object.keys(items).map<Item>(itemName => {
    const item = items[itemName];
    return fromItemWithName(itemName, item);
  });
};

type Search = (query: string) => Item[];
export const search: Search = query => {
  const items = getItemNames();
  const options: Fuse.IFuseOptions<string> = {
    minMatchCharLength: 2,
    threshold: 1,
  };
  const fuse = new Fuse(items, options);
  const results = fuse.search(query);
  return results.map(result => findById(result.item));
};
