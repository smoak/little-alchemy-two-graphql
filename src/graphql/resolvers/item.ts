import { GraphQLFieldResolver } from 'graphql';

import { ItemCombination, findAll, findById } from '../../data/repos/item';
import { Query } from '../types';

export const ItemsResolver: GraphQLFieldResolver<Query, unknown, unknown> = (_source, _args, _context, _info) =>
  findAll();

interface ItemWithNameResolverArgs {
  readonly name: string;
}
export const ItemWithNameResolver: GraphQLFieldResolver<Query, unknown, ItemWithNameResolverArgs> = (_source, args) =>
  findById(args.name);

export const SourceItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.source);

export const TargetItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.target);
