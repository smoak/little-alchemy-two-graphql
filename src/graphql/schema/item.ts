import {
  GraphQLBoolean,
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  Thunk,
} from 'graphql';
import { globalIdField } from 'graphql-relay';

import { Item, ItemCombination } from '../../data/repos/item';
import { SourceItemCombinationResolver, TargetItemCombinationResolver } from '../resolvers';

import { nodeInterface } from './node';

interface ItemFields extends GraphQLFieldConfigMap<Item, unknown, unknown> {
  readonly name: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly myths: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly id: GraphQLFieldConfig<unknown, unknown, unknown>;
  readonly creates: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly combinations: GraphQLFieldConfig<Item, unknown, unknown>;
}

interface ItemCombinationFields extends GraphQLFieldConfigMap<ItemCombination, unknown, unknown> {
  readonly source: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
  readonly target: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
}

type Fields = () => ItemFields;
const fields: Fields = () => ({
  name: { type: new GraphQLNonNull(GraphQLString) },
  myths: { type: new GraphQLNonNull(GraphQLBoolean) },
  id: globalIdField('Item', obj => obj.name),
  creates: {
    description: 'The items that this item creates',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
  combinations: {
    description: 'The items that create this item',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
});

const itemCombinationFields: Thunk<ItemCombinationFields> = () => ({
  source: { type: new GraphQLNonNull(ItemType), resolve: SourceItemCombinationResolver },
  target: { type: new GraphQLNonNull(ItemType), resolve: TargetItemCombinationResolver },
});

export const ItemType = new GraphQLObjectType<Item, unknown, unknown>({
  name: 'Item',
  description: 'An item',
  interfaces: [nodeInterface],
  fields,
  isTypeOf: source =>
    Object.keys(source).includes('name') &&
    Object.keys(source).includes('myths') &&
    Object.keys(source).includes('creates'),
});

export const ItemCombinationType = new GraphQLObjectType<ItemCombination, unknown, unknown>({
  name: 'ItemCombination',
  description: 'A combination of two items',
  fields: itemCombinationFields,
});
