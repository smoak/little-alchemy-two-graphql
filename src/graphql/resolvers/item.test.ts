jest.mock('../../data/repos/item');

import { ConnectionArguments } from 'graphql-relay';

import { Item, ItemCombination, findAll, findById } from '../../data/repos/item';
import { Query } from '../types';

import {
  itemCombinationConnectionResolver,
  itemWithNameResolver,
  itemsResolver,
  sourceItemCombinationResolver,
  targetItemCombinationResolver,
} from './item';

describe('item', () => {
  const items: Item[] = [{ combinations: [], creates: [], myths: false, name: 'test-item' }];
  const itemCombination: ItemCombination = {
    source: 'source-item',
    target: 'target-item',
  };

  beforeEach(() => {
    (findAll as jest.Mock).mockReturnValue(items);
    (findById as jest.Mock).mockReturnValue(items[0]);
  });

  describe('.itemsResolver', () => {
    let result: ReturnType<typeof itemsResolver>;

    beforeEach(() => {
      result = itemsResolver();
    });

    it('returns the expected items', () => {
      expect(result).toBe(items);
    });
  });

  describe('.itemWithNameResolver', () => {
    let result: ReturnType<typeof itemWithNameResolver>;

    beforeEach(() => {
      const query: Query = { items: [] };
      const args = { name: 'test-item' };

      result = itemWithNameResolver(query, args);
    });

    it('returns the expected item', () => {
      expect(result).toBe(items[0]);
    });
  });

  describe('.sourceItemCombinationResolver', () => {
    let result: ReturnType<typeof sourceItemCombinationResolver>;

    beforeEach(() => {
      result = sourceItemCombinationResolver(itemCombination);
    });

    it('returns the expected item', () => {
      expect(result).toBe(items[0]);
    });
  });

  describe('.targetItemCombinationResolver', () => {
    let result: ReturnType<typeof targetItemCombinationResolver>;

    beforeEach(() => {
      result = targetItemCombinationResolver(itemCombination);
    });

    it('returns the expected item', () => {
      expect(result).toBe(items[0]);
    });
  });

  describe('.itemCombinationConnectionResolver', () => {
    let result: ReturnType<typeof itemCombinationConnectionResolver>;

    beforeEach(() => {
      const item: Item = {
        name: 'test-item',
        combinations: [itemCombination],
        creates: [itemCombination],
        myths: false,
      };
      const args: ConnectionArguments = {};
      result = itemCombinationConnectionResolver(item, args);
    });

    it('returns the expected result', () => {
      expect(result).toMatchSnapshot();
    });
  });
});
