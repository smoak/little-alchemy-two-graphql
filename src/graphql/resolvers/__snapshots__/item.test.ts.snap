// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item .itemCombinationConnectionResolver returns the expected result 1`] = `
{
  "edges": [
    {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": {
        "source": "source-item",
        "target": "target-item",
      },
    },
  ],
  "pageInfo": {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjA=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`item .itemConnectionResolver returns the expected result 1`] = `
{
  "edges": [
    {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": {
        "combinations": [],
        "creates": [],
        "imageUrl": "https://picsum.photos/48",
        "myths": false,
        "name": "test-item",
      },
    },
  ],
  "pageInfo": {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjA=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`item .itemCreationConnectionResolver returns the expected result 1`] = `
{
  "edges": [
    {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": {
        "source": "source-item",
        "target": "target-item",
      },
    },
  ],
  "pageInfo": {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjA=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;
